File: common/src/main/java/dev/mayaqq/estrogen/registry/EstrogenBlockEntities.java
package dev.mayaqq.estrogen.registry;

import dev.mayaqq.estrogen.Estrogen;
import dev.mayaqq.estrogen.client.registry.blockRenderers.centrifuge.CentrifugeCogInstance;
import dev.mayaqq.estrogen.client.registry.blockRenderers.centrifuge.CentrifugeRenderer;
import dev.mayaqq.estrogen.client.registry.blockRenderers.cookieJar.CookieJarInstance;
import dev.mayaqq.estrogen.client.registry.blockRenderers.cookieJar.CookieJarRenderer;
import dev.mayaqq.estrogen.client.registry.blockRenderers.dreamBlock.DreamBlockInstance;
import dev.mayaqq.estrogen.client.registry.blockRenderers.dreamBlock.DreamBlockRenderer;
import dev.mayaqq.estrogen.registry.blockEntities.CentrifugeBlockEntity;
import dev.mayaqq.estrogen.registry.blockEntities.CookieJarBlockEntity;
import dev.mayaqq.estrogen.registry.blockEntities.DreamBlockEntity;
import uwu.serenity.critter.stdlib.blockEntities.BlockEntityEntry;
import uwu.serenity.critter.stdlib.blockEntities.BlockEntityRegistrar;

public class EstrogenBlockEntities {

    public static final BlockEntityRegistrar BLOCK_ENTITIES = BlockEntityRegistrar.create(Estrogen.MOD_ID);

    public static final BlockEntityEntry<CentrifugeBlockEntity> CENTRIFUGE = BLOCK_ENTITIES.entry("centrifuge", CentrifugeBlockEntity::new)
        .validBlock(EstrogenBlocks.CENTRIFUGE)
        .renderer(() -> CentrifugeRenderer::new)
        .transform(Transgenders.instance(() -> CentrifugeCogInstance::new, true))
        .register();

    public static final BlockEntityEntry<DreamBlockEntity> DREAM_BLOCK = BLOCK_ENTITIES.entry("dream_block", DreamBlockEntity::new)
        .validBlock(EstrogenBlocks.DREAM_BLOCK)
        .renderer(() -> DreamBlockRenderer::new)
        .transform(Transgenders.instanceController(() -> () -> DreamBlockInstance.CONTROLLER))
        .register();

    public static final BlockEntityEntry<CookieJarBlockEntity> COOKIE_JAR = BLOCK_ENTITIES.entry("cookie_jar", CookieJarBlockEntity::new)
        .validBlock(EstrogenBlocks.COOKIE_JAR)
        .renderer(() -> CookieJarRenderer::new)
        .transform(Transgenders.instance(() -> CookieJarInstance::new, false))
        .register();
}

File: src/main/kotlin/dev/mayaqq/estrogen/content/EstrogenBlockEntities.kt
package dev.mayaqq.estrogen.content

import dev.mayaqq.estrogen.MOD_ID
import dev.mayaqq.estrogen.client.content.blockRenderers.cookieJar.CookieJarRenderer
import dev.mayaqq.estrogen.client.content.blockRenderers.cookieJar.CookieJarVisual
import dev.mayaqq.estrogen.client.content.blockRenderers.dreamBlock.DreamBlockRenderer
import dev.mayaqq.estrogen.client.content.blockRenderers.dreamBlock.DreamBlockVisual
import dev.mayaqq.estrogen.client.content.blockRenderers.memorial.MemorialRenderer
import dev.mayaqq.estrogen.content.blockEntities.CookieJarBlockEntity
import dev.mayaqq.estrogen.content.blockEntities.DreamBlockEntity
import dev.mayaqq.estrogen.content.blockEntities.DreamCatcherBlockEntity
import dev.mayaqq.estrogen.content.blockEntities.MemorialBlockEntity
import invoke.kitty.kritter.registry.api.Registrar
import invoke.kitty.kritter.registry.api.entry.RegistryEntry
import invoke.kitty.kritter.registry.blockEntity.blockEntity
import invoke.kitty.kritter.registry.blockEntity.renderer
import net.minecraft.core.registries.Registries
import net.minecraft.world.level.block.entity.BlockEntityType

object EstrogenBlockEntities : Registrar<BlockEntityType<*>> by Registrar(MOD_ID, Registries.BLOCK_ENTITY_TYPE) {

    val DreamBlock: RegistryEntry<BlockEntityType<DreamBlockEntity>> = blockEntity("dream_block", ::DreamBlockEntity) {
        validBlock { EstrogenBlocks.DreamBlock.value!! }
        renderer(::DreamBlockRenderer)
        visual(::DreamBlockVisual) { false }
    }

    val CookieJar: RegistryEntry<BlockEntityType<CookieJarBlockEntity>> = blockEntity("cookie_jar", ::CookieJarBlockEntity) {
        validBlock { EstrogenBlocks.CookieJar.value!! }
        renderer(::CookieJarRenderer)
        visual(::CookieJarVisual) { false }
    }

    val DreamCatcher: RegistryEntry<BlockEntityType<DreamCatcherBlockEntity>> = blockEntity("dreamcatcher", ::DreamCatcherBlockEntity) {
        validBlock { EstrogenBlocks.DreamCatcher.value!! }
    }

    val Memorial: RegistryEntry<BlockEntityType<MemorialBlockEntity>> = blockEntity("memorial", ::MemorialBlockEntity) {
        validBlock { EstrogenBlocks.Memorial.value!! }
        renderer(::MemorialRenderer)
    }
}